include/envoy/router/router.h

```diff
@@ -1090,6 +1094,75 @@ class Config {
 };
 
 using ConfigConstSharedPtr = std::shared_ptr<const Config>;
+class GenericConnectionPoolCallbacks;
+class UpstreamRequest;
+class GenericUpstream;
+
+// An API for wrapping either an HTTP or a TCP connection pool.
+class GenericConnPool : public Logger::Loggable<Logger::Id::router> {
+public:
+  virtual ~GenericConnPool() = default;
+
+  // Called to create a new HTTP stream or TCP connection. The implementation
+  // is then responsible for calling either onPoolReady or onPoolFailure on the
+  // supplied GenericConnectionPoolCallbacks.
+  virtual void newStream(GenericConnectionPoolCallbacks* callbacks) PURE;
+  // Called to cancel a call to newStream. Returns true if a newStream request
+  // was canceled, false otherwise.
+  virtual bool cancelAnyPendingRequest() PURE;
+  // Optionally returns the protocol for the connection pool.
+  virtual absl::optional<Http::Protocol> protocol() const PURE;
+};
+
+// An API for the UpstreamRequest to get callbacks from either an HTTP or TCP
+// connection pool.
+class GenericConnectionPoolCallbacks {
+public:
+  virtual ~GenericConnectionPoolCallbacks() = default;
+
+  virtual void onPoolFailure(ConnectionPool::PoolFailureReason reason,
+                             absl::string_view transport_failure_reason,
+                             Upstream::HostDescriptionConstSharedPtr host) PURE;
+  virtual void onPoolReady(std::unique_ptr<GenericUpstream>&& upstream,
+                           Upstream::HostDescriptionConstSharedPtr host,
+                           const Network::Address::InstanceConstSharedPtr& upstream_local_address,
+                           const StreamInfo::StreamInfo& info) PURE;
+  virtual UpstreamRequest* upstreamRequest() PURE;
```

Comment:
Can you add comments for each of the API methods? It's not totally clear to me what this method does and why we need it vs getting passed the upstream in `onPoolReady`

Reply:
There's a bit of an ordering constraint here.

I wanted the onPoolReady to be able to pass the generic upstream, since the Tcp::onPoolReady knew to create a TcpUpstream, and the Http::onPoolReady would create an HttpUpstream.

Unfortunately both Tcp and Http classes need the upstream on creation.  I could TODO a refactor where they both get setUpstreamRequest shortly after creation and replace this call with the setter in GenericUpstream if you prefer?